Limit number of webhooks loaded in each request

thenbrent opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR proposes changing how webhooks are loaded on each request to:

1. only load those that are active; and
1. set a limit on the number loaded.

At the moment, all webhooks are loaded on all requests, regardless of how many there are, and whether those webhooks are disabled or paused. This can cause severe performance issues when there are large number of webhooks in the database. Even if those webhooks are disabled (or worse, corrupted data).

The limit can be increased on sites that actually have more than 100 active webhooks they need to run.

#### Note on Backward Compatibility

**Limiting the number of webhooks loaded could be considered a breaking change**. If a site has >100 webhooks, then some of those webhooks would cease to work after updating. To address that, the limit could be increased, or the limit could be removed until version 4.0.

Loading only active webhooks could also be a breaking change if a site was somehow relying on disabled or paused webhooks to still have `WC_Webhook::process()` attached to the hooks. The actual processing of the webhook won't differ, because the first thing `WC_Webhook::process()` does is check if the webhook is active before proceeding (by calling `WC_Webhook::should_deliver()`). However, because `WC_Webhook::enqueue()` will no longer be called for disabled and paused webhooks, `WC_Webhook::process()` will no longer be attached to hooks associated with the topic, so there is a slight change to public APIs. I feel like this is obscure enough that it should not prevent its inclusion in a minor release.

#### Backstory

I recently came across this issue on a site with 16,300 webhook rows in the database. Each row had just the default `WC_Webhook::$data` values (i.e. http://pic.pros.pr/e497b001a3c0 ). Meaning they all had the default status of `disabled` and an empty `name`, `delivery_url` and `topic`. That indicates they were probably created by some rogue infinite loop.

Despite that, all 16,300 webhooks were being initialised on every request. While it only took half a millisecond to load each webhook, loading 16,300 lead to a 5-10 second increase in page load times, as seen in this New Relic snapshot: http://pic.pros.pr/b55880d87b6b

This continued until the corrupted webhooks were manually removed from the database and `'woocommerce_webhook_ids'` transient cache. This New Relic screenshot does a good job showing the impact over a 46 hours period including the state before and after: http://pic.pros.pr/3a1822c40548

### How to test the changes in this Pull Request:

0. Using `master`
1. Generate 20,000 webhooks with not data other than status set to `disabled`
2. Load a page on your site
3. Observe the slow load time

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### Changelog entry

```
* Tweak - Load only active webhooks on each request to avoid performance issues loading large numbers of disabled webhooks.
* Tweak - Load a maximum of 100 webhooks per request by default to avoid performance issues on sites with an large set of invalid webhooks. 
```

thenbrent committed:
Add WC_Tests_Webhook_Functions::create_webhook()

To make it reusable.

thenbrent committed:
Add $limit param to wc_load_webhooks()

And the corresponding data stores. Defaults to null, meaning do not
limit, for backward compatibility.

thenbrent committed:
Only load 100 webhooks per request

To avoid slowing down page loads on sites with a large numbers of
webhooks.

thenbrent committed:
Add $status param to wc_load_webhooks()

And the corresponding data stores. Defaults to '', meaning do not
load only webhooks with a specific status. This maintains backward
compatibility. However, the call to wc_load_webhooks() within
WooCommerce::load_webhooks() can now only load active webhooks, as
they are the only ones that should be enqueued.

thenbrent committed:
Only load active webhooks on each request

To avoid slowing down page loads on sites with a large number of
disabled or paused webhooks, which do not need to be loaded or
enqueued.

thenbrent committed:
Use search_webhooks()

To avoid duplicate SQL

thenbrent committed:
PHPCS fixes

Add fixes required for Travis to pass the PR, but which aren't
related to the PR diff.

Commit:
Test against $wp_filter instead of has_filter()

Because we don't have the same WC_Webhook instance as used in
wc_load_webhooks(), so it's impossible to check if the same
object's process() method is attached as a callback.

Commit:
Fix test_wc_load_webhooks_status_and_limit()

As the status is being explicitly passed to wc_load_webhooks(), it
will load webhooks with that status, not just active.

NickGreen commented:
Just pinging this issue to let you know that we are seeing a number of customer tickets each week that this would likely solve.

rodrigoprimo reviewed (commented):
I did a first review of this PR, and I could confirm that currently all the webhooks are loaded on every page request and that negatively impacts the page load time and memory usage when the site has a few thousand webhooks.

I like the idea of loading only active webhooks by default. I'm not sure of the consequences of limiting the number of webhooks loaded (I will defer to @claudiosanches and @claudiulodro on this).

@thenbrent, I found some minor issues in this PR and left a few comments. Could you please take a look? Could you also fix the conflicts? Thanks.

thenbrent committed:
Remove unused $wpdb

thenbrent committed:
Clear webhook transients when status changes

Background discussion: https://github.com/woocommerce/woocommerce/pull/21427#discussion_r246881785

thenbrent commented on includes/data-stores/class-wc-webhook-data-store.php:
```diff
@@ -59,7 +59,7 @@ public function create( &$webhook ) {
 		$webhook->apply_changes();
 
-		delete_transient( 'woocommerce_webhook_ids' );
+		$this->delete_transients( $webhook->get_status( 'edit' ) );
```
Fixed with SHA: 2ed2572afeafbc

thenbrent commented:
Thanks for the review @rodrigoprimo. I've fixed up those issues. :)

thenbrent commented on includes/data-stores/class-wc-webhook-data-store.php:
```diff
@@ -200,18 +200,33 @@ public function get_api_version_number( $api_version ) {
 	 */
-	public function get_webhooks_ids() {
+	public function get_webhooks_ids( $status = '', $limit = null ) {
 		global $wpdb;
```
Fixed with SHA: 28ca9f1ec4a2cdf6e33e3f10caddbd5b457b1a00

rrennick commented on includes/data-stores/class-wc-webhook-data-store.php:
```diff
@@ -200,18 +200,33 @@ public function get_api_version_number( $api_version ) {
+			$ids = $this->search_webhooks(
+				array(
+					'limit'  => -1,
+					'status' => $status,
```
@thenbrent This works for me as described. 

However, the inactive webhooks are still loaded into the cache/transient. I added 32768 webhooks to my install and the transient value is 506787 bytes. Do the inactive webhooks need to be loaded into cache?

mikejolley commented:
@rodrigoprimo @thenbrent Pardon my ignorance, but could we have no limit, but load active webhooks only in 3.6? That would allow a limit to be set if needed (and a limit could be added in 4.0) but we'd improve matters by only loading active webhooks. I'm not sure what a valid use case would be to load disabled webhooks.

mikejolley commented:
@woocommerce/proton Changing this to 3.6 so it can be re-reviewed. Parts may be shippable earlier than 4.0.

mikejolley commented:
I've removed the limit in https://github.com/woocommerce/woocommerce/pull/22760 which should be safe for 3.6 🤞

timmyc commented:
Since this got closed w/out merge, should we remove it from the 3.6 milestone?

mikejolley commented:
@timmyc Yes - there was a replacement PR that was merged instead.
